### PR TITLE
Discovery token is filtered out in exception message [API-1738].

### PR DESCRIPTION
**Pull request by OzanCansel**
- Discovery token is filtered out at the source of exception.
- `token_should_not_be_leaked` test is added for this case.

### CODE REVIEW

File: hazelcast/test/src/HazelcastTests7.cpp [hunk 1]
```diff
@@ -1845,6 +1845,27 @@ TEST_F(cloud_discovery_test, invalid_token)
...
+    cloudConfig.discovery_token = "bwMqx9xp4RBkBWfdpE3WPsuCLd0CXNS";
+
+    auto discovery_token = cloudConfig.discovery_token;
+
+    try {
+        hazelcast::new_client(std::move(config)).get();
+        FAIL();
+    } catch (const exception::illegal_state& e) {
+        std::string message = e.what();
+        ASSERT_EQ(message.find(discovery_token), std::string::npos);
```
Comment (ihsandemir) [inline hazelcast/test/src/HazelcastTests7.cpp:1865]: you should also check the nested exception message!!!

File: hazelcast/src/hazelcast/client/spi.cpp [hunk 2]
```diff
@@ -3071,17 +3072,21 @@ std::unordered_map<address, address>
...
                                               std::string(CLOUD_URL_PATH) +
-                                                cloud_config_.discovery_token,
+                                                discovery_token,
                                               timeout_);
         auto& conn_stream = httpsConnection.connect_and_get_response();
         return parse_json_response(conn_stream);
     } catch (std::exception& e) {
+        std::string message{ e.what() };
+        boost::replace_all(message, discovery_token, "<DISCOVERY_TOKEN>");
         std::throw_with_nested(
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/spi.cpp:3087]: can it leak due to nested exception?

**Review by ihsandemir — CHANGES_REQUESTED** [review]

File: hazelcast/src/hazelcast/client/spi.cpp (hunk 2 above)
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/spi.cpp:3087]: On application crash it is not printed to `stdout`. But the token is included `nested_exception` that's why I added `filter` handler to `SyncHttpsRequest` earlier.

File: hazelcast/test/src/HazelcastTests7.cpp (hunk 1 above)
Comment (OzanCansel) [reply hazelcast/test/src/HazelcastTests7.cpp:1865]: Nested exception contains it.

Comment (OzanCansel) [reply hazelcast/test/src/HazelcastTests7.cpp]: Changed the implementation, `token` is filtered out in the class which is the source of the exception. So it is not leaked by neither inner nor outer exceptions.

There is a problem related with an unit test, I wrote the test to check inner exception message but `std::rethrow_if_nested` doesn't throw the inner exception. Added it to comment.

Comment (ihsandemir) [reply hazelcast/test/src/HazelcastTests7.cpp]: Let's öpen an issue for this.

File: hazelcast/include/hazelcast/util/SyncHttpsClient.h [hunk 3]
```diff
@@ -37,14 +37,16 @@ class HAZELCAST_API SyncHttpsClient
 public:
     SyncHttpsClient(const std::string& server_ip,
                     const std::string& uri_path,
-                    std::chrono::steady_clock::duration timeout);
+                    std::chrono::steady_clock::duration timeout,
+                    const std::string& secret_removal = std::string{});
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/util/SyncHttpsClient.h:41]: Reference to a temporary object???

Comment (OzanCansel) [reply hazelcast/include/hazelcast/util/SyncHttpsClient.h:41]: It is completely safe. [See ref.](https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary)

[An example](https://godbolt.org/z/zdrovvePT)

File: hazelcast/test/src/HazelcastTests7.cpp (hunk 1 above)
Comment (OzanCansel) [reply hazelcast/test/src/HazelcastTests7.cpp]: I will open after merge with the link of the test.

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**